source/common/common/cleanup.h

```diff
@@ -14,4 +15,18 @@ class Cleanup {
   std::function<void()> f_;
 };
 
+// RAII helper class to add an element to an std::list on construction and erase
+// it on destruction.
+template <class T> class ListAddAndRemove {
+public:
+  ListAddAndRemove(std::list<T>& container, T element) : container_(container) {
+    it_ = container.emplace(container.begin(), element);
+  }
+  virtual ~ListAddAndRemove() { container_.erase(it_); }
```

Comment:
Should this be a virtual class or final concrete? No strong opinions here, just curious to get your thoughts.

Reply:
If we mark it as final I don't think other classes can inherit from it. Or am I misunderstanding?